[Datasets] Add `size` parameter to `ImageFolderDatasource`

bveeramani opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

If you read a folder with differently-sized images, `ImageFolderDatasource` errors. This PR fixes the issue by resizing images to a user-specified size.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

ericl commented on python/ray/data/datasource/image_folder_datasource.py:
```diff
@@ -61,70 +60,60 @@ class ImageFolderDatasource(BinaryDatasource):
 
         image = iio.imread(data)
-        label = _get_class_from_path(path, self.root)
+        image = skimage.transform.resize(image, size)
```
Is this the most standard library for resizing images? Also, should we give options to customize the resizing process?

bveeramani replied:
> Is this the most standard library for resizing images?

I don't know if there's a de-facto standard. I chose scikit-image because it was easiest.

The three options are:
* PIL
* OpenCV
* scikit-image

PIL is clunky. To resize the array with PIL, we'd have to convert the array to a PIL image and then back to a NumPy array.
```
img = Image.fromarray(arr)
img.resize(size=(100, 100))
arr = np.array(img)
```

OpenCV is too big. According to Wikapedia, OpenCV is ~200MB.

bveeramani replied:
> Also, should we give options to customize the resizing process?

I think we should wait until someone asks.

For context, we could give any of the options exposed by scikit-image.

```
skimage.transform.resize(image, output_shape, order=None, mode='reflect', cval=0, clip=True, preserve_range=False, anti_aliasing=None, anti_aliasing_sigma=None)
```

TorchVision only exposes interpolation mode and anti-aliasing.

```
torchvision.transforms.Resize(size, interpolation=InterpolationMode.BILINEAR, max_size=None, antialias=None)
```

ericl replied:
Got it. Seems like we should revisit later with ray.data.read_images(), but this seems sensible for now.

amogkam commented on python/ray/data/tests/test_dataset_formats.py:
```diff
@@ -2802,7 +2794,9 @@ def test_image_folder_datasource_e2e(ray_start_regular_shared):
     from torchvision.models import resnet18
 
     root = os.path.join(os.path.dirname(__file__), "image-folder")
```
We should actually test whether the images have been resized right?

If the images in `image-folder` are all the same size to begin with, then this test will still pass even if resizing is not taking place.

bveeramani replied:
Good call. Fixed.

c21 commented on python/ray/data/datasource/image_folder_datasource.py:
```diff
@@ -133,6 +122,25 @@ def _read_file(self, f: "pyarrow.NativeFile", path: str, **reader_args):
             }
         )
 
+    def _check_import(self, *, module: str, package: str) -> None:
```
nice, shall we put this inside [data/_internal/util.py](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/util.py), so we can start use same method everywhere?

bveeramani replied:
Fixed

c21 commented on python/ray/data/datasource/image_folder_datasource.py:
```diff
@@ -61,70 +60,60 @@ class ImageFolderDatasource(BinaryDatasource):
-        partition_filter: PathPartitionFilter = None,
-        **kwargs,
+        root: str,
+        size: Tuple[int, int],
```
Do we want to do any sanity check of the user-provided `size`? What's the largest image size we want to support now?

bveeramani replied:
Added a check that the size contains positive numbers.

No idea what the large image size should be, if any.

c21 reviewed (approved):
LGTM

jiaodong reviewed (approved):
thanks for the quick fix ! One confirmation: i remember you previously mentioned our implementation assumes folder structure is same as imagenet, is this behavior still the same ? Looks like we're making it single root path now.

Also, how would user add / apply labels to each image read with `ImageFolderDatasource` API ?

bveeramani commented:
> thanks for the quick fix ! One confirmation: i remember you previously mentioned our implementation assumes folder structure is same as imagenet, is this behavior still the same ? Looks like we're making it single root path now.

Behavior is identical. Previously, `create_reader` would raise an error if `paths` contained more than one element. 

```
        if len(paths) != 1:
            raise ValueError(
                "`ImageFolderDatasource` expects 1 path representing the dataset "
                f"root, but it got {len(paths)} paths instead. To fix this "
                "error, pass in a single-element list containing the dataset root "
                '(for example, `paths=["s3://imagenet/train"]`)'
            )
```
To improve UX, I've renamed `paths` to `root` and changed the type to `str`.

> Also, how would user add / apply labels to each image read with `ImageFolderDatasource` API ?

By label, do you mean integer targets used for training?  This datasource already creates a "label" column.

c21 commented:
Looks like there's some test failure in AIR. Could you help fix these? Thanks.

```
==================== Test output for //doc/source/ray-air/examples:torch_image_batch_pretrained:
--
  | WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 2684354560 bytes available. This will harm performance! You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you can increase /dev/shm size by passing '--shm-size=9.34gb' to 'docker run' (or add it to the run_options list in a Ray cluster config). Make sure to set this to more than 30% of available RAM.
  | Started a local Ray instance. View the dashboard at http://127.0.0.1:8265.
  | Running GPU batch prediction with 1GB data from s3://anonymous@air-example-data-2/1G-image-data-synthetic-raw
  | Traceback (most recent call last):
  | File "/root/.cache/bazel/_bazel_root/5fe90af4e7d1ed9fcf52f59e39e126f5/execroot/com_github_ray_project_ray/bazel-out/k8-opt/bin/doc/source/ray-air/examples/torch_image_batch_pretrained.runfiles/com_github_ray_project_ray/doc/source/ray-air/examples/torch_image_batch_pretrained.py", line 33, in <module>
  | dataset = ray.data.read_datasource(ImageFolderDatasource(), paths=[data_url])
  | File "/ray/python/ray/data/read_api.py", line 274, in read_datasource
  | _wrap_and_register_arrow_serialization_workaround(read_args),
  | File "/ray/python/ray/_private/client_mode_hook.py", line 105, in wrapper
  | return func(*args, **kwargs)
  | File "/ray/python/ray/_private/worker.py", line 2239, in get
  | raise value.as_instanceof_cause()
  | ray.exceptions.RayTaskError(TypeError): ray::_get_read_tasks() (pid=3661, ip=172.16.16.3)
  | File "/ray/python/ray/data/read_api.py", line 1136, in _get_read_tasks
  | reader = ds.create_reader(**kwargs)
  | TypeError: create_reader() got an unexpected keyword argument 'paths'
  | ================================================================================
  | (13:28:10) [79 / 100] 14 / 21 tests, 1 failed; Testing //doc/source/ray-air/examples:torch_image_example; 199s local
  | (13:31:32) [80 / 101] 15 / 21 tests, 1 failed; Testing //doc/source/ray-air/examples:torch_incremental_learning; 43s local
  | (13:35:24) [80 / 101] 15 / 21 tests, 1 failed; Testing //doc/source/ray-air/examples:torch_incremental_learning; 275s local
  | (13:39:51) [85 / 106] 20 / 21 tests, 1 failed; Testing //doc/source/ray-air/examples:xgboost_starter; 10s local
```

bveeramani commented:
> Looks like there's some test failure in AIR. Could you help fix these? Thanks.

Okay, should be fixed.

richardliaw commented:
fix lint?

jiaodong reviewed (commented):
Thanks for updating our examples ! nit: You should be able to remove `transforms.Resize(256)` now given the resize is already done. Not a blocker.

richardliaw reviewed (approved):
stamp on docs
